pkg/controller/controller.go

```diff
@@ -0,0 +1,49 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	framework "github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework"
+	"time"
+)
+
+// ControllerPlugin is an interface to controller plugins that can be run
+// in kube-controller-manager.  A ControllerPluginManager handles instantiating
+// and managing the lifecycle of controllers.
+type ControllerPlugin interface {
+	// Init initializes the plugin.  This will be called once before Run.
+	Init(host ControllerHost)
+
+	// Name returns the plugin's name.  Plugins should use namespaced names
+	// such as "example.com/replication-controller".  The "kubernetes.io" namespace is
+	// reserved for plugins which are bundled with kubernetes.
+	Name() string
+
+	// DefaultSyncPeriod gets the suggested sync period for the plugin.
+	// The default period is not guaranteed to be honored.
+	DefaultSyncPeriod() time.Duration
+
+	// CreateController is a factory for creating controllers from this plugin
+	CreateController() Controller
+}
+
+// Controller represents a long-lived process that watches and synchronizes objects in the system.
+// Controllers watch any number of objects for changes, with one Config for each watch/process
+type Controller interface {
+	ControllerPlugin
```

Comment:
It's a little weird to embed the plugin into the instance.  For example, this is now legal:

``
plugin.CreateController().CreateController().CreateController().Init(host)
```

If you want some of the same methods, define them explicitly.